### PR TITLE
Integrating MaAsLin2 into Galaxy

### PR DESCRIPTION
**Pull request by EngyNasr**
Adding [**MaAsLin2**](http://huttenhower.sph.harvard.edu/maaslin) tool: which is comprehensive R package for efficiently determining multivariable association between phenotypes, environments, exposures, covariates and microbial meta’omic features. MaAsLin2 relies on general linear models to accommodate most modern epidemiological study designs, including cross-sectional and longitudinal, and offers a variety of data exploration, normalization, and transformation methods.

This wrapper includes: 1 tool xml file, 1 macros, 1 yml file and the test data folder. 

I need guidance in the boolean arguments, which is not a big problem but better if improved, the false value is set to FALSE however when the button is set to no (in the local use galaxy) the false value is not given to the command.  Any advice ?
I handled that with filtering the output later on in the code.

Thanks a lot,
Engy :)

### CODE REVIEW

**Comment by EngyNasr** [comment]
Doing: Reducing the test data size and updating tests accordingly

File: tools/maaslin2/maaslin2.xml [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+                <option value="CPLM">CPLM</option>
+                <option value="NEGBIN">NEGBIN</option>
+                <option value="ZINB">ZINB</option>
+            </param>
+            <param argument="--correction" type="text" value="BH" optional="true" label="Correction" help="The correction method for computing  the q-value"/>
+            <param argument="--standardize" type="boolean" truevalue="TRUE" falsevalue="FALSE" checked="true" label="Apply z-score so continuous metadata are on  the same scale"/>
+            <param argument="--cores" type="integer" value="1" optional="true" label="Cores" help="The number of R processes to run in parallel"/>        
+        </section>          
+        <section name="output" title="Set Plotting Output" expanded="true">
+            <param argument="--plot_heatmap" type="boolean" truevalue="TRUE" falsevalue="FALSE" checked="true" label="Generate a heatmap for the significant  associations"/>
```
Comment (bgruening) [inline tools/maaslin2/maaslin2.xml:95]: How does the commandline need to look like? I assume simply `--plot_heatmap` and not `--plot_heatmap TRUE`?

If so try this.
```suggestion
            <param argument="--plot_heatmap" type="boolean" truevalue="--plot_heatmap" falsevalue="" checked="true" label="Generate a heatmap for the significant  associations"/>
```

File: tools/maaslin2/maaslin2.xml [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+    --fixed_effects '$fixed_effects'
+#end if
+#if $additional_options.correction
+    --correction '$additional_options.correction'
+#end if
+#if $additional_options.standardize
+    --standardize '$additional_options.standardize'
+#end if
+#if $output.plot_heatmap
+    --plot_heatmap '$output.plot_heatmap'
```
Comment (bgruening) [inline tools/maaslin2/maaslin2.xml:45]: Replace the entire if with "$output.plot_heatmap" if you can use my suggestion below.

File: tools/maaslin2/maaslin2.xml [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+        <section name="output" title="Set Plotting Output" expanded="true">
+            <param argument="--plot_heatmap" type="boolean" truevalue="TRUE" falsevalue="FALSE" checked="true" label="Generate a heatmap for the significant  associations"/>
+            <param argument="--heatmap_first_n" type="integer" value="50" optional="true" label="Heatmap plot first N" help="In heatmap, plot top N features with significant associations"/>
+            <param argument="--plot_scatter" type="boolean" truevalue="TRUE" falsevalue="FALSE" checked="true" label="Generate scatter plots for the significant associations"/>    
+        </section>   
+   </inputs>
+   <outputs>    
+        <data name="all_results" format="tsv" from_work_dir="outputFolder/all_results.tsv" label="All results ordered by increasing q-value"/>
+        <data name="significant_results" format="tsv" from_work_dir="outputFolder/significant_results.tsv" label="Q-values smaller than or equal to the threshold"/>
+        <data name="residuals" format="rdata" from_work_dir="outputFolder/residuals.rds" label="Data frame with residuals for each feature"/>
```
Comment (bgruening) [inline tools/maaslin2/maaslin2.xml:103]: is this output needed? Should it be optional?

File: tools/maaslin2/maaslin2.xml [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+   <outputs>    
+        <data name="all_results" format="tsv" from_work_dir="outputFolder/all_results.tsv" label="All results ordered by increasing q-value"/>
+        <data name="significant_results" format="tsv" from_work_dir="outputFolder/significant_results.tsv" label="Q-values smaller than or equal to the threshold"/>
+        <data name="residuals" format="rdata" from_work_dir="outputFolder/residuals.rds" label="Data frame with residuals for each feature"/>       
+        <data format="pdf" name="headmap" from_work_dir="outputFolder/figures/heatmap.pdf" label="Heatmap of the significant associations" >
+            <filter>(output['plot_heatmap'] is True)</filter>
+        </data>        
+        <collection name="figures_pdfs" type="list" label="Plots" >
+            <discover_datasets pattern="(?P&lt;designation&gt;.+)" directory="outputFolder/figures/" format="pdf"/>
+            <filter>(output['plot_scatter'] is True)</filter>
```
Comment (bgruening) [inline tools/maaslin2/maaslin2.xml:109]: the brackets are not needed imho

File: tools/maaslin2/maaslin2.xml [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+    --cores '$additional_options.cores'
+#end if
+    'input_data.tsv'
+    'input_metadata.tsv'
+    'outputFolder'
+&&
+cd outputFolder &&  mkdir -p figures/ && cp *.pdf figures
+    ]]></command>
+    <inputs>
+        <param name="input_data" type="data" format="tsv" label="Data (or features) file"/>
```
Comment (bgruening) [inline tools/maaslin2/maaslin2.xml:63]: please use `tabular`, `tsv` should probably be never used

File: tools/maaslin2/maaslin2.xml [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+        <param name="input_metadata" type="data" format="tsv" label="Metadata file"/>
+        <param argument="--fixed_effects" type="text" value="diagnosis,dysbiosisnonIBD,dysbiosisUC,dysbiosisCD,antibiotics,age" multiple="true" optional="true" label="Interactions: Fixed effects" help="The fixed effects for the model,  comma-delimited for multiple effects"/>
```
Comment (bgruening) [inline tools/maaslin2/maaslin2.xml:65]: from where does the user get this list? Do we need to add a validator here?

**Review by bgruening — COMMENTED** [review]
Thanks @EngyNasr

File: tools/maaslin2/maaslin2.xml (hunk 5 above)
Comment (EngyNasr) [reply tools/maaslin2/maaslin2.xml:63]: sorrryy

Comment (EngyNasr) [reply tools/maaslin2/maaslin2.xml:63]: I will change it directly

File: tools/maaslin2/maaslin2.xml (hunk 6 above)
Comment (EngyNasr) [reply tools/maaslin2/maaslin2.xml:65]: supposedly, the default value of this argument is ALL so I checked their documentation to know what is this ALL and wrote all options they support such that user by default will have all the list and have the option to remove from it

File: tools/maaslin2/maaslin2.xml (hunk 3 above)
Comment (EngyNasr) [reply tools/maaslin2/maaslin2.xml:103]: it is outputted in all ways, there is no tool's argument that makes it optional, but I can do it

File: tools/maaslin2/maaslin2.xml (hunk 1 above)
Comment (bgruening) [reply tools/maaslin2/maaslin2.xml:95]: In that case you can use `truevalue="--plot_heatmap TRUE"`

File: tools/maaslin2/maaslin2.xml (hunk 3 above)
Comment (bgruening) [reply tools/maaslin2/maaslin2.xml]: Yes, this is usually the case. But the Galaxy tool dev can decide if its a good idea to expose all those files by default to the user or not.

File: tools/maaslin2/maaslin2.xml (hunk 6 above)
Comment (bgruening) [reply tools/maaslin2/maaslin2.xml:65]: if this is a fixed set of values and no free text, why not offering a select field here?

File: tools/maaslin2/maaslin2.xml (hunk 1 above)
Comment (EngyNasr) [reply tools/maaslin2/maaslin2.xml:95]: I will do this but that is actually not the only problem, the main problem is when I turn it off, the argument is not given as False, else its just not written in the command, and that for the tool means true and not false

Comment (EngyNasr) [reply tools/maaslin2/maaslin2.xml:95]: that solved the problem, thanks a lot :))

**Comment by EngyNasr** [comment]
Can I create a PR now for [step 3](https://galaxyproject.eu/posts/2020/08/22/three-steps-to-galaxify-your-tool/) 🎉 ?

**Comment by bgruening** [comment]
> Can I create a PR now for [step 3](https://galaxyproject.eu/posts/2020/08/22/three-steps-to-galaxify-your-tool/) tada ?

only when this one is merged :) otherwise your step 3 will fail :)

File: tools/maaslin2/macros.xml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+            <edam_topic>topic_3305</edam_topic>
+        </edam_topics>           
+    </xml>
+    <xml name="requirements">
+        <requirements>
+            <requirement type="package" version="@TOOL_VERSION@">maaslin2</requirement>
+            <yield/>
+        </requirements>
+    </xml>
+    <token name="@HELP_HEADER@">
```
Comment (bgruening) [inline tools/maaslin2/macros.xml:19]: isn't that redundant with what you have in the other help section?

Comment (EngyNasr) [reply tools/maaslin2/macros.xml:19]: I will remove it

**Review by bgruening — APPROVED** [review]

**Comment by bgruening** [comment]
Thanks @EngyNasr! Now its time for step 3.

**Pull request merged**